File: Python3/Medium/0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        left = root
        right = root
        leftHeight = 0
        rightHeight = 0

        while left:
            leftHeight += 1
            left = left.left
        
        while right:
            rightHeight += 1
            right = right.right

        if leftHeight == rightHeight:
            return pow(2, leftHeight) - 1
        
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)

File: C/Medium/0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.c
#include <math.h>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int countNodes(struct TreeNode* root) {
    if (root == NULL) return 0;

    struct TreeNode* left = root;
    struct TreeNode* right = root;
    int leftHeight = 0;
    int rightHeight = 0;

    while (left != NULL) {
        leftHeight++;
        left = left->left;
    }

    while (right != NULL) {
        rightHeight++;
        right = right->right;
    }

    // 'root' is a complete tree
    if (leftHeight == rightHeight) {
        return pow(2, leftHeight) - 1;
    }
    
    return 1 + countNodes(root->left) + countNodes(root->right);
}